### ISSUE TITLE
Add revision to static asset 

### ISSUE THREAD

**Issue by gpedro** [yeoman/generator-webapp#255]
#### References
- [gulp-rev](https://github.com/sindresorhus/gulp-rev)

**Comment by silvenon** [comment]
There is a [recipe](https://github.com/yeoman/generator-gulp-webapp/blob/b4878dc1306d52ba58b0226e8a9870600004209b/docs/recipes/asset-revisioning.md) for this, but I think you're right, maybe we should implement this.

**Comment by gpedro** [comment]
in generator(grunt)-webapp it's by default
https://github.com/yeoman/generator-webapp/blob/master/app/templates/Gruntfile.js#L255-L268

**Comment by silvenon** [comment]
Yeah, it would be nice if it was implemented here, too. The main problem is that I haven't figured out a way to rev all the assets, not just CSS & JS :disappointed: Otherwise I would implement it right away.

**Comment by hivre** [comment]
I'm working on a solution using `gulp-rev-all` (see: https://github.com/smysnk/gulp-rev-all)
It looks like it can rev all assets in a smart manner.

I'll make a pull-request once I have it successfully working on my local install, should be around this weekend

**Comment by silvenon** [comment]
Great! :+1: I tried, but no success.

**Comment by hivre** [comment]
I have a first-setup working, but there's a bug that rev-all can't rewrite references in minified html files. Also, the current setup leaves revved and unrevved files in the dist (the latter should get removed)

A simple solution would be to move the html-minify line to a separate task, and use something like gulp-napkin to remove unrevved files. 

But what I'd like to propose is a little bit more involved than the most simple solution:

* split the html task in a dedicated html task and a separate minify task and minify-html task
* the html-task at this moment would be rather empty (but you could add html-hint etc. it also makes extending it easier)
* the minify task will do all (and only) minification, except html files. 
* write all minified files to `/.tmp`
* `build` reads `/.tmp` and revs it, then minifies html files.
 
I prefer this as it separates `.tmp` and `dist` more, thus not needing an additional gulp-plugin (aside from the extendability that granular separate tasks provide). and the naming of the html-task is off, it only does minification atm.
But I don't want to hijack this either, and there are probably a few other solutions available.

**Comment by silvenon** [comment]
Could you post an example of how would that `html` task look like?

**Comment by hivre** [comment]
hmm, strike that task. 
I personally use it for aria and html linting, but I see that's my own addition...

So:
* split the html task in ~~a dedicated html task and~~ a separate minify task and minify-html task

**Comment by silvenon** [comment]
If this would help revving, I'm all for it. But I think it's best to perform it on the `dist` folder after the build, that way we could rev fonts and images too. Or does splitting the `html` task make that easier/possible?

**Comment by hivre** [comment]
My vision is:

`start build -> copy*/minify() -> dest(.tmp) -> rev(.tmp) -> dest(dist) -> minifyHtml(dist)`

That way the dist always has the latest and greatest, ready for deployment. The .tmp is a temporary dir where we collect everything before we rev it and copy it to dist.
But I am also happy to see if gulp-napkin works as expected and use that to clean up the non-revved copies of files in dist, and leave the current situation as is.

* files we can't minify should just get copied.

**Comment by hivre** [comment]
pull-request #293 uses the short-solution:

It uses `gulp-rev-napkin` to get rid of duplicate, un-revved files, and makes little changes to the gulp-file.

I have never done any testing -- I know, rather shamefull -- but I'm using this on my own project, and it works.

**Comment by sondr3** [comment]
Have you tried using the `gulp-rev-replace` plugin instead of `gulp-rev-napkin`? Just curious since the latter hasn't been updated in a while.

**Comment by hivre** [comment]
wasn't aware of `gulp-rev-replace`. 
I'll try it out asap.

**Comment by hivre** [comment]
I tried out `gulp-rev-replace`, but can't get it to work.

Tbh. I can see `rev-napkin` not being updated if it works well. it's only 111 lines of code...

**Comment by hivre** [comment]
I've got something with rev-replace running, but it actually removes the revved files, and leaves the un-revved files alone. Exactly the opposite we want :)

Might be my current implementation, I'm not using a manifest file. When I did that, I got an `undefined` error, most likely because the file doesn't exist (is in the stream) when it's used. Probably can get around this by creating a separate task, like in the rev-replace example. But that's precisely what I'm trying not to do.

**Comment by silvenon** [comment]
> but it actually removes the revved files, and leaves the un-revved files alone

I was using it before and it worked fine. Maybe it gets screwed up when not replacing in the same stream as the one where revving is done?

**Comment by sondr3** [comment]
I've been using `rev-replace` for a while in my optimizing task and it does exactly what it should, removes the unrevved files and inserts references to the revved ones in the HTML. It should work.

**Comment by hivre** [comment]
@silvenon  that's actually what I'm doing :) All in the same stream...
@sondr3 rev-all already replaces the references. I use napkin to remove un-revved files only.

**Comment by sondr3** [comment]
But that's not really a problem if you properly do the tasks that will fix the assets, I never have non-revved assets in my dist-folder so `gulp-napkin` shouldn't be necessary at all. You need to take a look at the other tasks and maybe have them move their assets to a `.tmp` folder and move the revved files into the `dist` folder, that way you have no need for `gulp-napkin`.

Also, I thought that `gulp-rev` just renames the files itself but doesn't rename the references in the HTML, in other words you'll end up with the CSS still being named `main.css` while the one that you want is actually `main-0b819as.css`. Have you tried live serving your site after you've removed the non-revved files? AFAIK it should try to serve assets that do not exist since they reference the old files.

**Comment by hivre** [comment]
But that's not really a problem if you properly do the tasks that will fix the assets, 
    I never have non-revved assets in my dist-folder so gulp-napkin shouldn't be necessary at all. 
    You need to take a look at the other tasks and maybe have them move their assets to a .tmp folder and move the revved files into the dist folder, that way you have no need for gulp-napkin.

I will friendly refer you to my earlier message of 4 hours ago, which literally mentions this in the last sentence, or the message of 5 days ago where I mention this as well.

    Also, I thought that gulp-rev just renames the files itself but doesn't rename the references in the HTML, in other words you'll end up with the CSS still being named main.css while the one that you want is actually main-0b819as.css. 

I'm not using `gulp-rev`, but `gulp-rev-all`.  gulp-rev-all takes child-references into account for its hashing, therefore it's better than gulp-rev (and it saves us from  the dependency on `gulp-rev-replace`, which is a nice bonus.)

    Have you tried live serving your site after you've removed the non-revved files? AFAIK it should try to serve assets that do not exist since they reference the old files.

I can recommend that you try it as well, you'll see how remarkably well it works :smile:

**Comment by sondr3** [comment]
Yes, I do realise that, I started writing my comment as I was reading and didn't spot it before I reached the end. I am also using `gulp-rev-all` but whenever I disable `gulp-rev-replace` it doesn't rewrite the references, I just tried and without using `gulp-rev-replace` none of the references were replaced. Weird, I never knew `gulp-rev-all` was supposed to do that, I'm going to take a look again. Sorry for being a bother :smiley:

**Comment by hivre** [comment]
I'm sorry for the non-response. I'm picking this up today

**Comment by illycz** [comment]
Some news about that?

**Comment by silvenon** [comment]
@illycz no. 😕 We should really solve this, but Gulp isn't a good tool for this task.

**Comment by yuvashrikarunakaran** [comment]
nstall the Plugin: First, install gulp-rev by running the following command:

bash
Copy code
npm install --save-dev gulp-rev
Basic Example: Below is a simple example of how to use gulp-rev to append content hashes to your static assets, such as CSS and JavaScript files.

js
Copy code
const gulp = require('gulp');
const rev = require('gulp-rev');

gulp.task('revision', function () {
  return gulp.src('src/assets/**/*.{css,js}')
    .pipe(rev())  // Append content hash
    .pipe(gulp.dest('dist/assets'))  // Save the revised files to dist
    .pipe(rev.manifest())  // Create a manifest file
    .pipe(gulp.dest('dist/assets'));  // Save the manifest to dist
});
How It Works:

gulp-rev generates unique file names based on their contents, like style.css becoming style-2d0d54d8.css.
The rev.manifest() generates a JSON file (rev-manifest.json) mapping the original filenames to the hashed filenames for reference.
Using Manifest: After generating the manifest, you might want to rewrite the paths to these revised assets in your HTML. For that, you can use gulp-rev-replace.

First, install it:

bash
Copy code
npm install --save-dev gulp-rev-replace
Then, replace the references in your HTML files like so:

js
Copy code
const revReplace = require('gulp-rev-replace');

gulp.task('update-html', function () {
  const manifest = gulp.src('dist/assets/rev-manifest.json');

  return gulp.src('src/*.html')
    .pipe(revReplace({ manifest }))
    .pipe(gulp.dest('dist'));
});
This will update the asset references in your HTML to point to the newly revisioned files (e.g., style-2d0d54d8.css instead of style.css).

Key Features:
Cache Busting: It helps prevent browser caching issues by changing the file names whenever their content changes.
Manifest File: Keeps track of the original file names and the new revisioned file names, allowing easy path updating in your HTML or other files.
Compatibility: Works well with other Gulp plugins, like gulp-rev-replace for automatically updating references in HTML.
Full Example Workflow:
Here’s an example combining multiple tasks including revisioning:

js
Copy code
const gulp = require('gulp');
const rev = require('gulp-rev');
const revReplace = require('gulp-rev-replace');
const useref = require('gulp-useref');
const uglify = require('gulp-uglify');
const cssnano = require('gulp-cssnano');

gulp.task('useref', function () {
  return gulp.src('src/*.html')
    .pipe(useref())  // Concatenate CSS and JS files
    .pipe(gulp.dest('dist'));
});

gulp.task('revision', function () {
  return gulp.src('dist/assets/**/*.{css,js}')
    .pipe(rev())
    .pipe(gulp.dest('dist/assets'))
    .pipe(rev.manifest())
    .pipe(gulp.dest('dist/assets'));
});

gulp.task('revreplace', function () {
  const manifest = gulp.src('dist/assets/rev-manifest.json');

  return gulp.src('dist/*.html')
    .pipe(revReplace({ manifest }))
    .pipe(gulp.dest('dist'));
});

gulp.task('build', gulp.series('useref', 'revision', 'revreplace'));
